Add documentation for running models in Python scripts

lbianchi-lbl opened the pull request:
## Fixes: #431 

## Summary/Motivation:

- Document recommended structure for Python scripts to avoid errors on Windows such as e.g. #387 

## Changes proposed in this PR:

- Add how-to guide on running models in Python scripts

### Legal Acknowledgement

By contributing to this software project, I agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the license terms described in the LICENSE.txt file at the top level of this directory.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

lbianchi-lbl committed:
Add documentation for running models in Python scripts

lbianchi-lbl commented:
Currently, for the example code [here](https://watertap--453.org.readthedocs.build/en/453/how_to_guides/how_to_run_models_in_a_py_script.html#example-python-file-with-recommended-structure), I mostly used the [snippet](https://github.com/aladshaw3/watertap/blob/dd81ed04d00c108664892cae807aa0b7fd28e5fb/basic_test.py) that @aladshaw3 was using in #390 to investigate the error, but if you can think of a simple way to make the example code in more representative (e.g. actually initialize/solve the model?), I'd be happy to add it.

aladshaw3 committed:
Merge branch 'main' into add-docs-py-scripts

aladshaw3 committed:
Merge branch 'main' into add-docs-py-scripts

aladshaw3 commented on docs/how_to_guides/how_to_run_models_in_a_py_script.rst:
```diff
@@ -0,0 +1,113 @@
+   m.o = pyo.Objective(expr=m.x)
+
+   nlp = PyomoNLP(m)
+
+
```
```suggestion


Example: Python file with recommended structure to run RO model
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

.. code-block::

   # Import concrete model from Pyomo
   from pyomo.environ import ConcreteModel
   # Import flowsheet block from IDAES core
   from idaes.core import FlowsheetBlock
   # Import NaCl property model
   import watertap.property_models.NaCl_prop_pack as props
   # Import utility tool for calculating scaling factors
   from idaes.core.util.scaling import calculate_scaling_factors
   # Import RO model
   from watertap.unit_models.reverse_osmosis_0D import ReverseOsmosis0
   # import the solver 
   from idaes.core.util import get_solver
   
   # Get default watertap solver 
   solver = get_solver()

   # Put all the model constructors, initialization, and solver in a separate function
   def main():

       # Create a concrete model, flowsheet, and NaCl property parameter block.
       m = ConcreteModel()
       m.fs = FlowsheetBlock(default={"dynamic": False})
       m.fs.properties = props.NaClParameterBlock()

       # Add an RO unit to the flowsheet.
       m.fs.unit = ReverseOsmosis0D(default={"property_package": m.fs.properties})
       
       # Specify system variables.
       m.fs.unit.inlet.flow_mass_phase_comp[0, 'Liq', 'NaCl'].fix(0.035)  # mass flow rate of NaCl
       m.fs.unit.inlet.flow_mass_phase_comp[0, 'Liq', 'H2O'].fix(0.965)  # mass flow rate of water
       m.fs.unit.inlet.pressure[0].fix(50e5)  # feed pressure
       m.fs.unit.inlet.temperature[0].fix(298.15)  # feed temperature
       m.fs.unit.area.fix(50)  # membrane area
       m.fs.unit.A_comp.fix(4.2e-12)  # membrane water permeability
       m.fs.unit.B_comp.fix(3.5e-8)  # membrane salt permeability
       m.fs.unit.permeate.pressure[0].fix(101325)  # permeate pressure
       
       # Set scaling factors for component mass flowrates.
       m.fs.properties.set_default_scaling('flow_mass_phase_comp', 1, index=('Liq', 'H2O'))
       m.fs.properties.set_default_scaling('flow_mass_phase_comp', 1e2, index=('Liq', 'NaCl'))
       
       # Calculate scaling factors.
       calculate_scaling_factors(m)
       
       # Initialize the model passing default solver options 
       m.fs.unit.initialize(optarg=solver.options)
       
       # Solve the model (using the tee=True option to display solver info)
       results = solver.solve(m, tee=True)

   # Call that function in the "__main__" for the script 
   if __name__ == "__main__":
       main()
```

lbianchi-lbl replied:
Ooh, I really like this. I'll add this as-is as the "good example", and the same code without the `main()` function for the "bad example".

aladshaw3 reviewed (changes requested):
I think we can keep the simple examples you have and just add another simple example of the 'proper' way to write the script. Here, I show the proper format for the RO unit.

lbianchi-lbl committed:
Use more realistic code for examples

lbianchi-lbl commented:
@aladshaw3 I've changed the examples to use the code you suggested.

The only change I made is moving the call to `get_solver()` to within the `main()` function (since that's what the rest of the guide prescribes); I'm not sure it's strictly required (i.e. whether it would cause the same sort of errors on Windows as #387), but it's more consistent. Let me know if you can think of any drawbacks for this.

**EDIT** I'm sorry, I've read your review comment more carefully, and realized you were suggesting to have the more realistic example *in addition* to the simple/trivial examples, whereas I've changed to using it *instead* of the simple examples. Let me know if you think the current (i.e. c431b1b) is still OK, otherwise I'll add the simpler examples back.

aladshaw3 reviewed (approved):
Current version (i.e. https://github.com/watertap-org/watertap/commit/c431b1ba129a20eb39e03deefd8937bb8d86cd06) is fine to me. So long as it shows a good and bad example for this, I think it is fine. 

You may need to 'kick' github actions to re-run the checks. It seems like it is stuck waiting for something.

lbianchi-lbl committed:
Trigger CI

ksbeattie reviewed (approved):
LGTM
